File: mobile/cloudity_shared/lib/calendar_repeat.dart
import 'cloudity_datetime.dart';

/// Récurrence agenda — même sémantique que `tasks.repeat_rule`.
class CalendarRepeatOption {
  const CalendarRepeatOption(this.value, this.label);
  final String value;
  final String label;
}

const kCalendarRepeatOptions = <CalendarRepeatOption>[
  CalendarRepeatOption('', 'Pas de répétition'),
  CalendarRepeatOption('daily', 'Chaque jour'),
  CalendarRepeatOption('weekdays', 'Jours ouvrés (lun–ven)'),
  CalendarRepeatOption('weekly', 'Chaque semaine'),
  CalendarRepeatOption('monthly', 'Chaque mois'),
];

const kNoteColorOptions = <({String id, String label})>[
  (id: 'default', label: 'Défaut'),
  (id: 'yellow', label: 'Jaune'),
  (id: 'green', label: 'Vert'),
  (id: 'blue', label: 'Bleu'),
  (id: 'pink', label: 'Rose'),
  (id: 'purple', label: 'Violet'),
  (id: 'orange', label: 'Orange'),
  (id: 'teal', label: 'Sarcelle'),
  (id: 'red', label: 'Rouge'),
  (id: 'gray', label: 'Gris'),
];

String? normalizeCalendarRepeat(String? raw) {
  final s = raw?.trim();
  if (s == null || s.isEmpty) return null;
  if (s == 'daily' || s == 'weekdays' || s == 'weekly' || s == 'monthly') {
    return s;
  }
  return null;
}

DateTime _addOccurrence(DateTime start, String rule) {
  switch (rule) {
    case 'daily':
      return start.add(const Duration(days: 1));
    case 'weekdays':
      var n = start.add(const Duration(days: 1));
      while (n.weekday == DateTime.saturday || n.weekday == DateTime.sunday) {
        n = n.add(const Duration(days: 1));
      }
      return n;
    case 'weekly':
      return start.add(const Duration(days: 7));
    case 'monthly':
      return DateTime(
        start.year,
        start.month + 1,
        start.day,
        start.hour,
        start.minute,
        start.second,
        start.millisecond,
      );
    default:
      return start.add(const Duration(days: 1));
  }
}

/// Déplie les événements récurrents dans [rangeStart, rangeEnd).
List<Map<String, dynamic>> expandCalendarEvents(
  List<Map<String, dynamic>> events, {
  required DateTime rangeStart,
  required DateTime rangeEnd,
}) {
  final out = <Map<String, dynamic>>[];
  const maxOcc = 400;

  for (final ev in events) {
    final rule = normalizeCalendarRepeat(ev['repeat_rule']?.toString());
    final startRaw = (ev['start_at'] ?? ev['starts_at'])?.toString();
    final endRaw = (ev['end_at'] ?? ev['ends_at'])?.toString();
    final baseStart = parseCloudityDateTime(startRaw);
    final baseEnd = parseCloudityDateTime(endRaw);
    if (baseStart == null || baseEnd == null) continue;
    final duration = baseEnd.difference(baseStart);
    if (duration.isNegative) continue;

    if (rule == null) {
      if (baseStart.isBefore(rangeEnd) && baseEnd.isAfter(rangeStart)) {
        out.add(Map<String, dynamic>.from(ev));
      }
      continue;
    }

    var cursor = baseStart;
    var guard = 0;
    while (!cursor.add(duration).isAfter(rangeStart) && guard < 2000) {
      cursor = _addOccurrence(cursor, rule);
      guard++;
    }

    var n = 0;
    while (cursor.isBefore(rangeEnd) && n < maxOcc) {
      final occEnd = cursor.add(duration);
      if (cursor.isBefore(rangeEnd) && occEnd.isAfter(rangeStart)) {
        final copy = Map<String, dynamic>.from(ev);
        copy['start_at'] = cursor.toUtc().toIso8601String();
        copy['end_at'] = occEnd.toUtc().toIso8601String();
        copy['occurrence_key'] = '${ev['id']}:${cursor.toUtc().toIso8601String()}';
        out.add(copy);
      }
      cursor = _addOccurrence(cursor, rule);
      n++;
    }
  }

  out.sort((a, b) {
    final sa = parseCloudityDateTime((a['start_at'] ?? '').toString()) ?? DateTime(0);
    final sb = parseCloudityDateTime((b['start_at'] ?? '').toString()) ?? DateTime(0);
    return sa.compareTo(sb);
  });
  return out;
}

File: frontend/apps/cloudity-web/src/lib/calendarRepeat.ts
/** Récurrence agenda — alignée sur tasks.repeat_rule (pas de RRULE RFC). */

export type CalendarRepeatRule = 'daily' | 'weekdays' | 'weekly' | 'monthly'

export const CALENDAR_REPEAT_OPTIONS: { value: '' | CalendarRepeatRule; label: string }[] = [
  { value: '', label: 'Pas de répétition' },
  { value: 'daily', label: 'Chaque jour' },
  { value: 'weekdays', label: 'Jours ouvrés (lun–ven)' },
  { value: 'weekly', label: 'Chaque semaine' },
  { value: 'monthly', label: 'Chaque mois' },
]

export function normalizeCalendarRepeat(raw?: string | null): CalendarRepeatRule | null {
  if (!raw) return null
  if (raw === 'daily' || raw === 'weekdays' || raw === 'weekly' || raw === 'monthly') return raw
  return null
}

function addOccurrence(start: Date, rule: CalendarRepeatRule): Date {
  const n = new Date(start.getTime())
  switch (rule) {
    case 'daily':
      n.setDate(n.getDate() + 1)
      return n
    case 'weekdays': {
      do {
        n.setDate(n.getDate() + 1)
      } while (n.getDay() === 0 || n.getDay() === 6)
      return n
    }
    case 'weekly':
      n.setDate(n.getDate() + 7)
      return n
    case 'monthly':
      n.setMonth(n.getMonth() + 1)
      return n
  }
}

export type ExpandedCalendarEvent<T extends { id: number; start_at: string; end_at: string; repeat_rule?: string | null }> =
  T & {
    /** Début de l’occurrence affichée (ISO). */
    occurrence_key: string
  }

/**
 * Déplie les événements récurrents dans [rangeStart, rangeEnd).
 * Les occurrences virtuelles partagent le même `id` (édition/suppression = série).
 */
export function expandCalendarEvents<
  T extends { id: number; start_at: string; end_at: string; repeat_rule?: string | null },
>(events: T[], rangeStart: Date, rangeEnd: Date): ExpandedCalendarEvent<T>[] {
  const out: ExpandedCalendarEvent<T>[] = []
  const maxOcc = 400

  for (const ev of events) {
    const rule = normalizeCalendarRepeat(ev.repeat_rule)
    const baseStart = new Date(ev.start_at)
    const baseEnd = new Date(ev.end_at)
    if (Number.isNaN(baseStart.getTime()) || Number.isNaN(baseEnd.getTime())) continue
    const durationMs = Math.max(0, baseEnd.getTime() - baseStart.getTime())

    if (!rule) {
      if (baseStart < rangeEnd && baseEnd > rangeStart) {
        out.push({ ...ev, occurrence_key: `${ev.id}:${baseStart.toISOString()}` })
      }
      continue
    }

    let cursor = new Date(baseStart.getTime())
    // Avancer jusqu’à entrer dans la fenêtre
    let guard = 0
    while (cursor.getTime() + durationMs <= rangeStart.getTime() && guard < 2000) {
      cursor = addOccurrence(cursor, rule)
      guard++
    }

    let n = 0
    while (cursor < rangeEnd && n < maxOcc) {
      const occEnd = new Date(cursor.getTime() + durationMs)
      if (cursor < rangeEnd && occEnd > rangeStart) {
        out.push({
          ...ev,
          start_at: cursor.toISOString(),
          end_at: occEnd.toISOString(),
          occurrence_key: `${ev.id}:${cursor.toISOString()}`,
        })
      }
      cursor = addOccurrence(cursor, rule)
      n++
    }
  }

  return out.sort((a, b) => new Date(a.start_at).getTime() - new Date(b.start_at).getTime())
}
